SYCL/Basic/device-selectors.cpp

```diff
@@ -0,0 +1,88 @@
+// RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple %s -o %t.out
+// RUN:  %t.out
+
+#include <CL/sycl.hpp>
+using namespace sycl;
+
+auto exception_handler_lambda = [](exception_list elist) {
+  // nop
+};
+
+void handle_exceptions_f(exception_list elist) {
+  // nop
+}
+
+auto refuse_any_device_lambda = [](const device &d) { return -1; };
+
+int refuse_any_device_f(const device &d) { return -1; }
+
+int main() {
+  // Ensure these compile correctly. That exception handler does not
+  // end up specialized against device selector.
+  queue exception_queue_01(exception_handler_lambda);
+  queue exception_queue_02(handle_exceptions_f);
+
+  // Set up callable device selector.
+  auto deviceList = device::get_devices();
+  device lastDevice = deviceList.back();
+  auto select_last_device = [&lastDevice](const device &d) {
+    return d == lastDevice;
+  };
+
+  // Instantiate platform via callable Device Selector.
+  platform lastPlatform(select_last_device);
+
+  // Test each of the four queue constructors that take callable Device
+  // Selectors: q(device selector) q(dev selector ,  async handler) q(context ,
+  // dev selector) q(context, dev selector , async handler)
+  queue lastQueue(select_last_device);
+  assert(lastQueue.get_device().get_platform() == lastPlatform &&
+         "queue and platform created by selecting same device, should result "
+         "in matching platforms");
+
+  queue lastQueueWithHandler(select_last_device, exception_handler_lambda);
+  assert(lastQueueWithHandler.get_device() == lastDevice &&
+         "queue created by selecting a device should have that same device");
+
+  context ctx = lastQueue.get_context();
```

Comment:
What are the devices in the `ctx` at this point? Everything from last device's platform?

Reply:
who knows? It's most likely a context of one device, but it could be every device in the platform, or the main device plus a bunch of sub-devices. At least in theory.

Reply:
If it's one device only, then we're not testing much here...

Reply:
I'd rather do

```
{
  context ctx(platform.get_all_devices()); // or whatever.
  queue q(ctx, ...);
}
````

Reply:
done.  It's a bit more complicated than that, but I agree that it's best to try to make the test not-so-trivial.